File: php/src/Node.php
<?php


namespace Tree;


class Node
{
    /*** @var mixed */
    public $value;
    public ?Node $right = null;
    public ?Node $left = null;

    public function __construct($value = null)
    {
        $this->value = $value;
    }

    public function insertRight($value): Node
    {
        return $this->right = new Node($value);
    }

    public function insertLeft($value): Node
    {
        return $this->left = new Node($value);
    }

    public function preorderTraversal(): array
    {
        $numbers = [$this->value];

        if ($this->left !== null) {
            $numbers = array_merge($numbers, $this->left->preorderTraversal());
        }

        if ($this->right !== null) {
            $numbers = array_merge($numbers, $this->right->preorderTraversal());
        }

        return $numbers;
    }

    public function inorderTraversal(): array
    {
        $numbers = [];

        if ($this->left !== null) {
            $numbers = array_merge($numbers, $this->left->inorderTraversal());
        }

        $numbers[] = $this->value;

        if ($this->right !== null) {
            $numbers = array_merge($numbers, $this->right->inorderTraversal());
        }

        return $numbers;
    }

    public function postorderTraversal(): array
    {
        $numbers = [];

        if ($this->left !== null) {
            $numbers = array_merge($numbers, $this->left->postorderTraversal());
        }

        if ($this->right !== null) {
            $numbers = array_merge($numbers, $this->right->postorderTraversal());
        }

        $numbers[] = $this->value;
        return $numbers;
    }
}

File: cpp/src/Node.cpp
#include <memory>
#include "include/Node.hpp"

Node::Node(int value) : value(value), right(nullptr), left(nullptr)
{
}

std::shared_ptr<Node> Node::insertRight(int v)
{
    auto node = std::make_shared<Node>(v);
    right = node;
    return node;
}

std::shared_ptr<Node> Node::insertLeft(int v)
{
    left = std::make_shared<Node>(v);
    return left;
}

std::list<int> Node::preorderTraversal() const
{
    std::list<int> list;

    list.push_back(value);

    if (left != nullptr)
    {
        std::list<int> leftValues = left->preorderTraversal();
        list.merge(leftValues);
    }

    if (right != nullptr)
    {
        std::list<int> rightValues = right->preorderTraversal();
        for (int i: rightValues)
        {
            list.push_back(i);
        }
    }
    return list;
}

std::list<int> Node::inorderTraversal() const
{
    std::list<int> list;

    if (left != nullptr)
    {
        std::list<int> leftValues = left->inorderTraversal();
        list.merge(leftValues);
    }

    list.push_back(value);

    if (right != nullptr)
    {
        std::list<int> rightValues = right->inorderTraversal();
        for (int i: rightValues)
        {
            list.push_back(i);
        }
    }
    return list;
}

std::list<int> Node::postorderTraversal() const
{
    std::list<int> list;

    if (left != nullptr)
    {
        std::list<int> leftValues = left->postorderTraversal();
        list.merge(leftValues);
    }


    if (right != nullptr)
    {
        std::list<int> rightValues = right->postorderTraversal();
        for (int i: rightValues)
        {
            list.push_back(i);
        }
    }
    list.push_back(value);
    return list;
}
